### PR TITLE
15642 - Updating Add to channel modal

### PR DESCRIPTION
**Pull request by asaadmahmood**
#### Summary
15642 - Updating Add to channel modal

#### Ticket Link
https://github.com/mattermost/mattermost-server/issues/15642

#### Screenshots
Screenshot 2020-12-06 at 6 40 19 PM
Screenshot 2020-12-06 at 6 40 26 PM

### CODE REVIEW

File: components/channel_invite_modal/channel_invite_modal.tsx
Comment (jespino) [inline components/channel_invite_modal/channel_invite_modal.tsx:286]: This is not a good change, this is going to break how you can "translate the string", in english, clearly you said "Add people to Whatever", but maybe in another language you would say "XXXX ChannelName XXXX" so... you can't assume the order of the words in an translations.

Comment (asaadmahmood) [reply components/channel_invite_modal/channel_invite_modal.tsx:286]: @jespino That's the way it was.
I just changed New Members to people.

Previously it was:
"channel_invite.addNewMembers": "Add New Members to ",

I made it:
 "channel_invite.addNewMembers": "Add people to ",

Nothing else.

Comment (nevyangelova) [reply components/channel_invite_modal/channel_invite_modal.tsx:268]: Instead of using the <span> you can use the 'values' functionality of React Intl, and that would solve potential issues.

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [inline components/multiselect/multiselect.tsx:252]: what is the reasoning for these classNames? We wanted to migrate webapp styles to using BEM. More semantic classes would help with clarity.

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [inline components/multiselect/multiselect.tsx:266]: Let's not use inline styles as we are getting a lot of conflicts because of the existing ones.

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [inline components/multiselect/multiselect.tsx:335]: Can we build a props object and then use ternary statements for specific properties? I know it's more if's in practice but it will be more readable.

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [inline components/multiselect/multiselect.tsx:445]: same as above

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [inline components/multiselect/multiselect.tsx:504]: is it possible to use classNames for the extra classes instead of having this prop?

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [inline components/multiselect/multiselect.tsx:534]: nit: more semantic classes

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [inline components/multiselect/multiselect.tsx:563]: nit: no inline styles

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [inline components/multiselect/multiselect.tsx:562]: nit: more semantic classes

File: components/widgets/inputs/users_invite_input.jsx
Comment (nevyangelova) [inline components/widgets/inputs/users_invite_input.jsx:28]: Since we are migrating to typescript, all new files in webapp should be tsx.

File: components/widgets/inputs/users_invite_input.jsx
Comment (nevyangelova) [inline components/widgets/inputs/users_invite_input.jsx:113]: can we do this inline? I know this is an existent practice in mattermost but we have discussed this to be only for complex component changes.

File: components/widgets/inputs/users_invite_input.jsx
Comment (nevyangelova) [inline components/widgets/inputs/users_invite_input.jsx:159]: nit: semantic value naming.

File: components/widgets/inputs/users_invite_input.jsx
Comment (nevyangelova) [inline components/widgets/inputs/users_invite_input.jsx:219]: nit: more semantic variable name

File: sass/components/_channel-switcher.scss
Comment (nevyangelova) [inline sass/components/_channel-switcher.scss:88]: can we use the rgba variable here too?

File: sass/components/_channel-switcher.scss
Comment (nevyangelova) [inline sass/components/_channel-switcher.scss:127]: let's keep this style semantic as it's a problem within webapp we have constantly crashing styles because of direct element styling.

File: sass/components/_react-select.scss
Comment (nevyangelova) [inline sass/components/_react-select.scss:65]: can we use the rgba variables too? it's part of the migration to compass design system.

File: sass/utils/_modifiers.scss
Comment (nevyangelova) [inline sass/utils/_modifiers.scss:63]: nit: new line

**Review by nevyangelova — CHANGES_REQUESTED** [review]

**Comment by esethna** [comment]
Thanks @asaadmahmood, minor nit, I think we decided on the design pattern that the autocomplete would overlap the input fields slightly? See this ticket for example: https://mattermost.atlassian.net/browse/MM-30898

Looks like this may have also broken for the regular autocomplete as I believe it used to work that way? Would you be open to submitted a fix?
image

**Comment by asaadmahmood** [comment]
@esethna The design that we have in the Figma file does not have it overlapping the input.
I think the overlapping input was only for the post area input, and I think it would be better if we tackle that in a separate ticket.

Also can you point out what is broken because of this? I don't think we have the overlapping thing implemented yet.

image

**Comment by esethna** [comment]
@asaadmahmood it was implemented in the past as part of: https://mattermost.atlassian.net/browse/MM-22956 (but it is no longer overlapping on community or daily it seems)

I think we should be consistent here too, but I'll let @michaelgamble comment on that.

File: components/multiselect/multiselect.tsx
Comment (asaadmahmood) [reply components/multiselect/multiselect.tsx:252]: @nevyangelova These are just to avoid me creating another rule or add extra css class for the icon.
I can just add css to it, but when we just want to give basic css rules, we can just reuse the css we already have instead of adding classes for every simple thing right?

File: components/channel_invite_modal/channel_invite_modal.tsx
Comment (asaadmahmood) [reply components/channel_invite_modal/channel_invite_modal.tsx:268]: Did that.

**Review by nevyangelova — APPROVED** [review]

**Comment by matthewbirtch** [comment]
> @asaadmahmood it was implemented in the past as part of: https://mattermost.atlassian.net/browse/MM-22956 (but it is no longer overlapping on community or daily it seems)
> 
> I think we should be consistent here too, but I'll let @michaelgamble comment on that.

This PR contribution implements the autocomplete positioning update (not merged yet): https://github.com/mattermost/mattermost-webapp/pull/7082#issuecomment-740828480

**Comment by asaadmahmood** [comment]
@matthewbirtch So from what I see, we're not overlapping it vertically right?

**Comment by matthewbirtch** [comment]
To be honest, I kind of forgot about the input overlapping. That PR is more about the horizontal positioning.

I'd love to see it overlap the message input field, but we can do that as a later enhancement

**Comment by asaadmahmood** [comment]
Then do we overlap the suggestion list over the input in this ticket too, or keep it at the edge like the way it is?

**Comment by matthewbirtch** [comment]
> Then do we overlap the suggestion list over the input in this ticket too, or keep it at the edge like the way it is?

I don't think we can be consistent everywhere. For example, the search input uses an autocomplete popover, but it won't look right if the popover overlaps the input since it's a smaller input.

So, I say we leave it for now as-is (not overlapping) since it's not overlapping anywhere yet. We can look at whether we actually want to overlap the post draft input later in a separate PR

**Comment by asaadmahmood** [comment]
I agree. I don't think we can have them overlap everywhere.
@esethna So we will keep it as is for now, and the input overlapping can be done as a separate ticket as well as Matt pointed out.

**Comment by esethna** [comment]
@matthewbirtch @asaadmahmood works for me, can we open a follow-up ticket to look into that. Agree it may not be a global pattern in places like search, but with the larger input fields it could certainly make sense (ie post input, edit post, this modal etc)

**Review by nevyangelova — COMMENTED** [review]
Still some comments are unaddressed

**Comment by asaadmahmood** [comment]
@nevyangelova Which ones? Can you comment them again?

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [reply components/multiselect/multiselect.tsx:335]: @devinbinnie can you advice here?

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [reply components/multiselect/multiselect.tsx:504]: this comment ^

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [reply components/multiselect/multiselect.tsx:534]: ^ this one too

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [reply components/multiselect/multiselect.tsx:562]: here

File: components/multiselect/multiselect.tsx
Comment (nevyangelova) [reply components/multiselect/multiselect.tsx:563]: here

File: components/multiselect/multiselect.tsx
Comment (devinbinnie) [reply components/multiselect/multiselect.tsx:335]: Yeah I'd agree that we do this here, it will definitely be more readable.

Comment (asaadmahmood) [reply components/multiselect/multiselect.tsx:335]: @devinbinnie Is there a way to do this?
Not to give a prop at all conditionally?

It's not giving the `react-select` prop a different value based on a condition, rather its to either give a few props to it, or not at all.

File: components/multiselect/multiselect.tsx
Comment (asaadmahmood) [reply components/multiselect/multiselect.tsx]: Are we sure we want to do that, we would then have to create css rules for every simple thing we want to do if we want to be consistent I guess. So if I have to just give it a margin-left of 4px, I would have to create a rule for that in every separate instance instead of reusing a helper class right?

File: components/multiselect/multiselect.tsx
Comment (asaadmahmood) [reply components/multiselect/multiselect.tsx:504]: Not sure how we do this.
Do I add a className prop to the component itself and then use classNames in it?

Furthermore, we are also moving the saveButton itself in the DOM based on the prop value, rather than just changing the class.

Comment (asaadmahmood) [reply components/multiselect/multiselect.tsx:504]: Though I'm just removing the class altogether, and giving it through css.

File: components/multiselect/multiselect.tsx
Comment (devinbinnie) [reply components/multiselect/multiselect.tsx:335]: Yes, so instead of having 2 components that give different sets of props, you would specify each prop as such:
```
prop={booleanValue ? thisProp: thatProp}
```

**Comment by asaadmahmood** [comment]
@devinbinnie @nevyangelova Can we approve this PR, or talk about how we do the ternary properties for React Select?
This PR has been waiting for that quite a while.

**Review by esethna — APPROVED** [review]

**Comment by esethna** [comment]
@michaelgamble for a UX review

**Comment by esethna** [comment]
@asaadmahmood are we adding in the statuses to this PR as per conversation in https://github.com/mattermost/mattermost-webapp/pull/7194

We should match the style of the other autocomplete statuses I would think

**Comment by asaadmahmood** [comment]
@esethna These are done from my end, I'm waiting for @michaelgamble to review it.
I plan on adding the statuses in another PR, to avoid multiple features in a single PR. It allows us to every back then if there are issues without reverting the whole thing.

**Review by michaelgamble — APPROVED** [review]
Looks good to me :)

File: components/multiselect/multiselect.tsx
Comment (devinbinnie) [inline components/multiselect/multiselect.tsx:334]: Can we remove this comment?

**Review by devinbinnie — CHANGES_REQUESTED** [review]

**Review by devinbinnie — APPROVED** [review]

**Review by ogi-m — COMMENTED** [review]
Hey @asaadmahmood, a few issues/questions:
1. The autocomplete is always active instead of only when there's text input, so the "Add'' button is always overlapped and I can't add users to a channel.
2. The text color for selected user's names is hard to read on dark themes:
add_people

3. What's the expected text when there are no results? 
It says _No items found_, while the Invite Members To Team modal shows _No one found matching *entered text*_

**Comment by asaadmahmood** [comment]
1, 2, will fix that.
3. @michaelgamble Any thoughts on the text?

**Comment by asaadmahmood** [comment]
@ogi-m Fixed 1 and 2.
If those are fine, we can approve the PR.
I can update the text for 3 later when I get a confirmation from some of the designers.

**Review by josephbaylon — APPROVED** [review]
LGTM. Affected e2e tests were updated and passing locally. There's a minor text bug,

```
"multiselect.list.notFound": "No results found matching **{searchQuery}**"
```

displays as 
```
No results found matching ****
```

**Comment by ogi-m** [comment]
Retested issues 1 and 2 and they're fixed 👍

**Review by ogi-m — APPROVED** [review]

**Comment by metanerd** [comment]
Hi, this PR failed in master with a failing snapshot. Do we just have to update the snapshot or could there be something wrong?

**Comment by asaadmahmood** [comment]
@saturninoabril Thoughts?

**Comment by saturninoabril** [comment]
I think it just need to update the snapshot. It's a merge conflict from the other PR - https://github.com/mattermost/mattermost-webapp/pull/7338 - which added the new snapshot on the said failing component. PR submitted - https://github.com/mattermost/mattermost-webapp/pull/7387

**Pull request merged**
